lib/grant-types/authorization-code-grant-type.ts

```diff
@@ -0,0 +1,202 @@
+        'Invalid grant: authorization code is invalid',
+      );
+    }
+
+    if (!code.client) {
+      throw new ServerError(
+        'Server error: `getAuthorizationCode()` did not return a `client` object',
+      );
+    }
+
+    if (!code.user) {
+      throw new ServerError(
+        'Server error: `getAuthorizationCode()` did not return a `user` object',
+      );
+    }
+
+    if (code.client.id !== client.id) {
+      throw new InvalidGrantError(
+        'Invalid grant: authorization code is invalid',
+      );
+    }
+
+    if (!(code.expiresAt instanceof Date)) {
+      throw new ServerError(
+        'Server error: `expiresAt` must be a Date instance',
+      );
+    }
+
+    if (code.expiresAt.getTime() < Date.now()) {
+      throw new InvalidGrantError(
+        'Invalid grant: authorization code has expired',
+      );
+    }
+
+    if (code.redirectUri && !is.uri(code.redirectUri)) {
+      throw new InvalidGrantError(
+        'Invalid grant: `redirect_uri` is not a valid URI',
+      );
+    }
+
+    return code;
+  }
+
+  /**
+   * Validate the redirect URI.
+   *
+   * "The authorization server MUST ensure that the redirect_uri parameter is
+   * present if the redirect_uri parameter was included in the initial
+   * authorization request as described in Section 4.1.1, and if included
+   * ensure that their values are identical."
+   *
+   * @see https://tools.ietf.org/html/rfc6749#section-4.1.3
+   */
+
+  validateRedirectUri(request: Request, code: AuthorizationCode) {
+    if (!code.redirectUri) {
+      return;
+    }
+
+    const redirectUri = request.body.redirect_uri || request.query.redirect_uri;
+
+    if (!is.uri(redirectUri)) {
+      throw new InvalidRequestError(
+        'Invalid request: `redirect_uri` is not a valid URI',
+      );
+    }
+
+    if (redirectUri !== code.redirectUri) {
+      throw new InvalidRequestError(
+        'Invalid request: `redirect_uri` is invalid',
+      );
+    }
+  }
+
+  /**
+   * Revoke the authorization code.
+   *
+   * "The authorization code MUST expire shortly after it is issued to mitigate
+   * the risk of leaks. [...] If an authorization code is used more than once,
+   * the authorization server MUST deny the request."
+   *
+   * @see https://tools.ietf.org/html/rfc6749#section-4.1.2
+   */
+
+  async revokeAuthorizationCode(code: AuthorizationCode) {
+    const status = await this.model.revokeAuthorizationCode(code);
+    if (!status) {
+      throw new InvalidGrantError(
+        'Invalid grant: authorization code is invalid',
+      );
+    }
+
+    return code;
+  }
+
+  /**
+   * Save token.
+   */
+
+  async saveToken(
+    user: User,
+    client: Client,
+    authorizationCode: string,
+    scope: string,
+  ) {
+    const accessScope = await this.validateScope(user, client, scope);
+    const accessToken = await this.generateAccessToken(client, user, scope);
+    const refreshToken = await this.generateRefreshToken(client, user, scope);
```

Comment:
Technically they might resolve faster in parallel...

https://stackoverflow.com/questions/35612428/call-async-await-functions-in-parallel

Reply:
Might, but Promise.all breaks the processing when one of the method fails (in any order). This is a security relevant module and processing them one after the other is necessary, as e.g.. a validateScope fail would not result that a possible more cpu expensive accessToken generation would start or maybe as a side effect you generate an accessToken and refreshToken persisted in the backEnd and you could login even though the validation failed.

So in this case it is better to not run them all in parallel just because it will process faster.

Reply:
One valid question could have been, why we dont use the generated accessScope for the generateAccessToken- and generateRefreshToken-calls.